sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ReturnInFinallyCheck.kt

```diff
@@ -0,0 +1,119 @@
+import org.jetbrains.kotlin.psi.KtThrowExpression
+import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private class FinallyBlockVisitor(private val report: (KtElement) -> Unit) : KtTreeVisitorVoid() {
+    private var loopDepthCounter = 0
+    private var lambdaDepthCounter = 0
+    private val labelsInFinallyBlock = ArrayList<String>()
+    private val jumpsToLabel = ArrayList<KtExpressionWithLabel>()
+
+    override fun visitFinallySection(finallySection: KtFinallySection) {
+        super.visitFinallySection(finallySection)
+        jumpsToLabel
+            .filter { isEscapingTheBlock(it, labelsInFinallyBlock) }
+            .forEach(report)
+    }
+
+    override fun visitLabeledExpression(expression: KtLabeledExpression) {
+        expression.name?.let { labelsInFinallyBlock.add(it) }
+        super.visitLabeledExpression(expression)
+    }
+
+    override fun visitLambdaExpression(lambdaExpression: KtLambdaExpression) {
+        lambdaDepthCounter++
+        super.visitLambdaExpression(lambdaExpression)
+        lambdaDepthCounter--
+    }
+
+    override fun visitLoopExpression(loopExpression: KtLoopExpression) {
+        loopDepthCounter++
+        super.visitLoopExpression(loopExpression)
+        loopDepthCounter--
+    }
+
+    override fun visitBreakExpression(expression: KtBreakExpression) {
+        visitAndCheck(loopDepthCounter, expression)
+    }
+
+    override fun visitContinueExpression(expression: KtContinueExpression) {
+        visitAndCheck(loopDepthCounter, expression)
+    }
+
+    override fun visitReturnExpression(expression: KtReturnExpression) {
+        visitAndCheck(lambdaDepthCounter, expression)
+    }
+
+    override fun visitThrowExpression(expression: KtThrowExpression) {
+        report(expression)
+    }
+
+    private fun visitAndCheck(depthCounter: Int, expression: KtExpressionWithLabel) {
+        if (depthCounter == 0) {
+            report(expression)
+        } else if (expression.labelQualifier != null) {
+            jumpsToLabel.add(expression)
+        }
+        super.visitExpressionWithLabel(expression)
+    }
+}
+
+@Rule(key = "S1143")
+class ReturnInFinallyCheck : AbstractCheck() {
+
+    override fun visitFinallySection(finallySection: KtFinallySection, kotlinFileContext: KotlinFileContext) {
+        val visitor = FinallyBlockVisitor { kotlinFileContext.reportIssue(it, it.buildReportMessage()) }
+        finallySection.accept(visitor)
+    }
+}
+
+/**
+ * Checks if the label in the element points to an unexpected label
+ */
+private fun isEscapingTheBlock(element: KtExpressionWithLabel, expectedLabels: List<String>): Boolean {
+    // Because we only call this function on expressions that have a label, we can assume this is safe
+    val label = element.getLabelName()!!
+    return label !in expectedLabels
+}
```

Comment:
How about?

```suggestion
private fun isEscapingTheBlock(element: KtExpressionWithLabel, expectedLabels: List<String>): Boolean =
    // Because we only call this function on expressions that have a label, we can assume this is safe
    element.getLabelName()!! !in expectedLabels
```